res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js

```diff
@@ -0,0 +1,1336 @@
+  TraktorS2MK2.outputCallback(field.value, field.group, "!shift");
+}
+
+TraktorS2MK2.loadTrack = function(field) {
+  var splitted = field.id.split(".");
+  var group = splitted[0];
+  if (TraktorS2MK2.shift_pressed[group]) {
+    engine.setValue(field.group, "eject", field.value);
+  } else {
+    engine.setValue(field.group, "LoadSelectedTrack", field.value);
+  }
+}
+
+TraktorS2MK2.syncEnabled = function(field) {
+  var now = Date.now();
+
+  var splitted = field.id.split(".");
+  var group = splitted[0];
+  // If shifted, just toggle.
+  // TODO(later version): actually make this enable explicit master.
+  if (TraktorS2MK2.shift_pressed[group]) {
+    if (field.value === 0) {
+      return;
+    }
+    var synced = engine.getValue(field.group, "sync_enabled");
+    engine.setValue(field.group, "sync_enabled", !synced);
+  } else {
+    if (field.value === 1) {
+      TraktorS2MK2.sync_enabled_time[field.group] = now;
+      engine.setValue(field.group, "sync_enabled", 1);
+    } else {
+      var cur_enabled = engine.getValue(field.group, "sync_enabled");
+      if (!cur_enabled) {
+        // If disabled, and switching to disable... stay disabled.
+        engine.setValue(field.group, "sync_enabled", 0);
+        return;
+      }
+      // was enabled, and button has been let go.  maybe latch it.
+      if (now - TraktorS2MK2.sync_enabled_time[field.group] > 300) {
+        engine.setValue(field.group, "sync_enabled", 1);
+        return;
+      }
+      engine.setValue(field.group, "sync_enabled", 0);
+    }
+  }
+}
+
+TraktorS2MK2.cue = function(field) {
+  var splitted = field.id.split(".");
+  var group = splitted[0];
+  if (TraktorS2MK2.shift_pressed[group]) {
+    if (ShiftCueButtonAction == "REWIND") {
+      if (field.value === 0) {
+        return;
+      }
+      engine.setValue(field.group, "start_stop", 1);
+    } else if (ShiftCueButtonAction == "REVERSEROLL") {
+      engine.setValue(field.group, "reverseroll", field.value);
+    } else {
+      print ("Traktor S2 Mk2 WARNING: Invalid ShiftCueButtonAction picked.  Must be either REWIND " +
+           "or REVERSEROLL");
+    }
+  } else {
+    engine.setValue(field.group, "cue_default", field.value);
+  }
+}
+
+TraktorS2MK2.play = function(field) {
+  if (field.value === 0) {
+    return;
+  }
+  var splitted = field.id.split(".");
+  var group = splitted[0];
+  if (TraktorS2MK2.shift_pressed[group]) {
+    var locked = engine.getValue(field.group, "keylock");
+    engine.setValue(field.group, "keylock", !locked);
+  } else {
+    var playing = engine.getValue(field.group, "play");
+    var deckNumber = TraktorS2MK2.controller.resolveDeck(group);
+    // Failsafe to disable scratching in case the finishJogTouch timer has not executed yet
+    // after a backspin.
+    if (engine.isScratching(deckNumber)) {
+      engine.scratchDisable(deckNumber, false);
+    }
+    engine.setValue(field.group, "play", !playing);
+  }
+}
+
+TraktorS2MK2.jogTouch = function(field) {
```

Comment:
add a TODO to move this to a common library

Reply:
I'm not sure I agree with that.

Reply:
this is going to be the third controller script that copies and pastes this jogwheel handling code.  That sort of duplication is unnecessary and brittle

Reply:
I'm not investing more energy into this legacy API that we will be replacing soon.

Reply:
https://github.com/mixxxdj/mixxx/wiki/New-controller-system